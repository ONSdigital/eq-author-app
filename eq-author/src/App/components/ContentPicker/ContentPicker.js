import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { isNil } from "lodash";

import ButtonGroup from "components/Buttons/ButtonGroup";
import Button from "components/Buttons/Button";

import ContentPickerSingle from "App/components/ContentPicker/ContentPickerSingle";

const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
`;

const ActionButtons = styled(ButtonGroup)`
  padding: 1em 0;
  flex: 0 0 auto;
`;
export default class ContentPicker extends React.Component {
  static propTypes = {
    onSubmit: PropTypes.func,
    onClose: PropTypes.func,
    data: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired
      })
    ).isRequired,
    config: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        childKey: PropTypes.string
      })
    ).isRequired
  };

  state = {
    selectedItems: []
  };

  getDataAtLevel(level) {
    if (level === 0) {
      return this.props.data;
    }
    const parentLevel = level - 1;
    const parentItem = this.state.selectedItems[parentLevel];
    if (!parentItem) {
      return [];
    }
    const { config } = this.props;
    return parentItem[config[parentLevel].childKey];
  }

  handleTitleClick = level => {
    const { config } = this.props;
    this.setState(state => {
      let openLevel = level;
      if (state.openLevel === level && state.selectedItems[level]) {
        openLevel = Math.min(state.selectedItems.length, config.length - 1);
      }
      return {
        openLevel
      };
    });
  };

  handleOptionClick = (level, option) => {
    const { config } = this.props;
    this.setState(state => ({
      openLevel: Math.min(level + 1, config.length - 1),
      selectedItems: [...state.selectedItems.slice(0, level), option]
    }));
  };

  renderPickers() {
    const { config } = this.props;
    const { selectedItems, openLevel } = this.state;

    return config.map(({ id, title: defaultTitle }, level) => {
      const data = this.getDataAtLevel(level);

      const isDisabled = level > selectedItems.length || !data.length;
      const isHidden = level > openLevel;
      const isOpen = level === openLevel || config.length === 1;

      const selectedItem = selectedItems[level];
      const isLastLevel = config.length - 1 === level;
      const useCustomTitle =
        !isLastLevel && selectedItem && selectedItem.displayName;

      const title = useCustomTitle
        ? `${defaultTitle}: ${selectedItem.displayName}`
        : defaultTitle;

      return (
        <ContentPickerSingle
          key={id}
          title={title}
          data={data}
          disabled={isDisabled}
          hidden={isHidden}
          open={isOpen}
          selected={isOpen}
          selectedOption={(selectedItem || {}).id}
          onTitleClick={() => this.handleTitleClick(level)}
          onOptionClick={option => this.handleOptionClick(level, option)}
          childKey={config[level].childKey}
          data-test={`${id}-picker`}
        />
      );
    });
  }

  render() {
    const { config } = this.props;
    const { selectedItems } = this.state;
    const selectedItem = selectedItems[config.length - 1];
    return (
      <React.Fragment>
        <ContentWrapper>{this.renderPickers()}</ContentWrapper>
        <ActionButtons horizontal align="right">
          <Button
            variant="secondary"
            type="button"
            onClick={this.props.onClose}
            data-test="cancel-button"
          >
            Cancel
          </Button>
          <Button
            type="submit"
            variant="primary"
            disabled={isNil(selectedItem)}
            onClick={() => this.props.onSubmit(selectedItem)}
            data-test="submit-button"
          >
            Select
          </Button>
        </ActionButtons>
      </React.Fragment>
    );
  }
}
