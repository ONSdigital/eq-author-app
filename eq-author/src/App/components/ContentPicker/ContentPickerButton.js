import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Icon from "App/components/ContentPicker/icon-group.svg?inline";
import ContentPickerAccordionHeader from "App/components/ContentPicker/ContentPickerAccordionHeader";
import { colors } from "constants/theme";

const SelectedIcon = styled(Icon)`
  --color-text: ${colors.white};
  position: absolute;
  width: 2em;
  height: 2em;
  margin: auto;
  top: 0;
  left: 0;
  bottom: 0;
  pointer-events: none;
  path {
    fill: var(--color-text);
  }
`;

const ContentPickerButton = ({ hidden, selected, label, ...otherProps }) => {
  if (hidden) {
    return null;
  }

  return (
    <ContentPickerAccordionHeader selected={selected} {...otherProps}>
      {selected ? <SelectedIcon /> : null}
      {label}
    </ContentPickerAccordionHeader>
  );
};

ContentPickerButton.propTypes = {
  hidden: PropTypes.bool,
  selected: PropTypes.bool,
  label: PropTypes.string
};

export default ContentPickerButton;
