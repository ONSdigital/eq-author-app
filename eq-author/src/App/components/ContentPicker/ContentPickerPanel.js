import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import ScrollPane from "components/ScrollPane";
import { colors, radius } from "constants/theme";

const StyledScrollPane = styled(ScrollPane)`
  padding: 0.25em 0;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

const PanelWrapper = styled.div`
  padding: 0.6em 0.5em;
  position: relative;
  overflow: auto;
  border: 1px solid ${colors.lightGrey};
  border-top: none;
  border-radius: 0 0 ${radius} ${radius};
  display: flex;
  flex-grow: 1;
`;

const ContentPickerPanel = ({ id, open, labelledBy, children }) => {
  if (!open) {
    return null;
  }
  return (
    <PanelWrapper
      id={id}
      role="tabpanel"
      aria-hidden={!open}
      aria-labelledby={labelledBy}
    >
      <StyledScrollPane>{children}</StyledScrollPane>
    </PanelWrapper>
  );
};

ContentPickerPanel.propTypes = {
  id: PropTypes.string.isRequired,
  labelledBy: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired
};

export default ContentPickerPanel;
