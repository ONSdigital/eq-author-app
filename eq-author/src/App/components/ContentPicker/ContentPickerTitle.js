import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import chevronIcon from "App/components/ContentPicker/chevron.svg";
import ContentPickerAccordionHeader from "App/components/ContentPicker/ContentPickerAccordionHeader";

export const TitleButton = styled(ContentPickerAccordionHeader)`
  &::after {
    content: url(${chevronIcon});
    position: absolute;
    right: 1.5em;
    transform: rotate(270deg);
  }

  &[aria-expanded="false"]::after {
    transform: rotate(-90deg);
    transition: transform 150ms ease-out;
  }

  &[aria-expanded="true"] {
    border-radius: 5px 5px 0 0;
  }
  ::after {
    transform: rotate(0deg);
    transition: transform 150ms ease-in;
  }
`;

class ContentPickerTitle extends React.Component {
  static propTypes = {
    open: PropTypes.bool.isRequired,
    children: PropTypes.node.isRequired
  };

  render() {
    const { children, open, ...otherProps } = this.props;

    return (
      <TitleButton
        role="tab"
        aria-expanded={open}
        aria-selected={open}
        {...otherProps}
      >
        {children}
      </TitleButton>
    );
  }
}

export default ContentPickerTitle;
