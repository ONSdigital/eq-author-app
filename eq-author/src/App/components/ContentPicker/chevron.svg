<svg width="12px" height="9px" viewBox="0 0 12 9" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">

    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Artboard" fill-rule="nonzero" fill="#FFFFFF">
            <polygon id="Shape" points="1.41 0.84 6 5.42 10.59 0.84 12 2.25 6 8.25 0 2.25"></polygon>
        </g>
    </g>
</svg>
