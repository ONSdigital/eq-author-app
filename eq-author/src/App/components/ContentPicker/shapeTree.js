import { find, isArray, pick, reduce } from "lodash";

const shapeTree = answers =>
  reduce(
    answers,
    (result, value) => {
      const answer = pick(value, ["id", "displayName"]);
      const page = pick(value.page, ["id", "displayName"]);
      const section = pick(value.page.section, ["id", "displayName"]);

      if (!find(result, { id: section.id })) {
        result.push(section);
      }

      const currentSection = find(result, { id: section.id });

      if (!find(currentSection.pages, { id: page.id })) {
        if (!isArray(currentSection.pages)) {
          currentSection.pages = [];
        }
        currentSection.pages.push(page);
      }

      const currentPage = find(currentSection.pages, { id: page.id });

      if (!find(currentPage.answers, { id: answer.id })) {
        if (!isArray(currentPage.answers)) {
          currentPage.answers = [];
        }
        currentPage.answers.push(answer);
      }

      return result;
    },
    []
  );

export const shapePageTree = answers =>
  reduce(
    answers,
    (result, value) => {
      const page = pick(value, ["id", "displayName"]);
      const section = pick(value.section, ["id", "displayName"]);

      if (!find(result, { id: section.id })) {
        result.push(section);
      }

      const currentSection = find(result, { id: section.id });

      if (!find(currentSection.pages, { id: page.id })) {
        if (!isArray(currentSection.pages)) {
          currentSection.pages = [];
        }
        currentSection.pages.push(page);
      }

      return result;
    },
    []
  );

export default shapeTree;
