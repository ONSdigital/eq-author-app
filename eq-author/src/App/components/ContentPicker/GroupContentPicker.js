import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { isNil } from "lodash";

import ButtonGroup from "components/Buttons/ButtonGroup";
import Button from "components/Buttons/Button";

import ContentPickerSingle from "App/components/ContentPicker/ContentPickerSingle";
import ContentPickerButton from "App/components/ContentPicker/ContentPickerButton";

const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
`;

const ActionButtons = styled(ButtonGroup)`
  padding: 1em 0;
  flex: 0 0 auto;
`;

export default class GroupContentPicker extends React.Component {
  static propTypes = {
    onSubmit: PropTypes.func,
    onClose: PropTypes.func,
    data: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
    config: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        childKey: PropTypes.string,
        groupKey: PropTypes.string
      })
    ).isRequired
  };

  state = {
    openLevel: null,
    selectedItem: null
  };

  handleTitleClick = level => {
    this.setState(state => {
      let openLevel = level;
      if (state.openLevel === level) {
        openLevel = null;
      }
      return {
        openLevel,
        selectedItem: null
      };
    });
  };

  handleOptionClick = (level, option) => {
    const config = this.props.config[level];

    this.setState({
      openLevel: level,
      selectedItem: {
        ...option,
        config
      }
    });
  };

  renderPickers() {
    const { config } = this.props;
    const { selectedItem, openLevel } = this.state;

    return config.map(({ id, title, expandable }, level) => {
      const isOpen = level === openLevel;
      const isHidden = openLevel !== null ? level > openLevel : false;
      const data = this.props.data[this.props.config[level].groupKey];
      const isDisabled = !data.length;

      if (expandable) {
        return (
          <ContentPickerSingle
            key={id}
            title={title}
            data={data}
            hidden={isHidden}
            open={isOpen}
            selected={isOpen}
            selectedOption={(selectedItem || {}).id}
            onTitleClick={() => this.handleTitleClick(level)}
            onOptionClick={option => this.handleOptionClick(level, option)}
            childKey={config[level].childKey}
            data-test={`${id}-picker`}
            disabled={isDisabled}
          />
        );
      } else {
        return (
          <ContentPickerButton
            key={id}
            onClick={() => this.handleOptionClick(level, { id, title })}
            selected={isOpen}
            open={isOpen}
            hidden={isHidden}
            label={title}
            data-test={`${id}-picker`}
            disabled={isDisabled}
          />
        );
      }
    });
  }

  render() {
    const { selectedItem } = this.state;
    return (
      <React.Fragment>
        <ContentWrapper>{this.renderPickers()}</ContentWrapper>
        <ActionButtons horizontal align="right">
          <Button
            variant="secondary"
            type="button"
            onClick={this.props.onClose}
            data-test="cancel-button"
          >
            Cancel
          </Button>
          <Button
            type="submit"
            variant="primary"
            disabled={isNil(selectedItem)}
            onClick={() => this.props.onSubmit(selectedItem)}
            data-test="submit-button"
          >
            Select
          </Button>
        </ActionButtons>
      </React.Fragment>
    );
  }
}
