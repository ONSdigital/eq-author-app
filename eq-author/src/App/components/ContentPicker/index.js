import React from "react";
import ContentPicker from "App/components/ContentPicker/ContentPicker";
import GroupContentPicker from "App/components/ContentPicker/GroupContentPicker";

const config = [
  {
    id: "section",
    title: "Section",
    childKey: "pages"
  },
  {
    id: "page",
    title: "Question",
    childKey: "answers"
  },
  {
    id: "answer",
    title: "Answer"
  }
];

export const AnswerContentPicker = props => (
  <ContentPicker {...props} config={config} />
);

const metadataConfig = [
  {
    id: "metadata",
    title: "Metadata"
  }
];

export const MetadataContentPicker = props => (
  <ContentPicker {...props} config={metadataConfig} />
);

const questionConfig = [
  {
    id: "section",
    title: "Section",
    childKey: "pages"
  },
  {
    id: "page",
    title: "Question"
  }
];

export const QuestionContentPicker = props => (
  <ContentPicker {...props} config={questionConfig} />
);

const routingDestinationConfig = [
  {
    id: "pages",
    title: "Other pages in this section",
    groupKey: "questionPages",
    expandable: true,
    destination: {
      absoluteDestination: {
        destinationId: null,
        destinationType: "QuestionPage"
      }
    }
  },
  {
    id: "sections",
    title: "Other sections",
    groupKey: "sections",
    expandable: true,
    destination: {
      absoluteDestination: {
        destinationId: null,
        destinationType: "Section"
      }
    }
  },
  {
    id: "endOfQuestionnaire",
    title: "End of questionnaire",
    groupKey: "logicalDestinations",
    expandable: false,
    type: "RoutingLogicalDestination",
    destination: {
      logicalDestination: {
        destinationType: "EndOfQuestionnaire"
      }
    }
  }
];

export const RoutingDestinationContentPicker = props => (
  <GroupContentPicker {...props} config={routingDestinationConfig} />
);
