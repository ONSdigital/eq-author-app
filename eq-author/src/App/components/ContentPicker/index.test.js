import React from "react";
import { shallow } from "enzyme";

import { AnswerContentPicker } from "App/components/ContentPicker";

describe("Answer Content Picker", () => {
  it("should render a configured ContentPicker", () => {
    expect(shallow(<AnswerContentPicker data={[]} />)).toMatchSnapshot();
  });
});
