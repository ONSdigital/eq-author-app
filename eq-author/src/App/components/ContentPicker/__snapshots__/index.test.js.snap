// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer Content Picker should render a configured ContentPicker 1`] = `
<ContentPicker
  config={
    Array [
      Object {
        "childKey": "pages",
        "id": "section",
        "title": "Section",
      },
      Object {
        "childKey": "answers",
        "id": "page",
        "title": "Question",
      },
      Object {
        "id": "answer",
        "title": "Answer",
      },
    ]
  }
  data={Array []}
/>
`;
